runtime/sys.c

```diff
@@ -191,6 +191,7 @@ CAML_RUNTIME_EVENTS_DESTROY();
 #ifndef NATIVE_CODE
   caml_debugger(PROGRAM_EXIT, Val_unit);
 #endif
+  caml_terminate_all_domains();
   if (caml_params->cleanup_on_exit)
     caml_shutdown();
```

Comment:
Do we want to call `caml_terminate_all_domains` unconditionally or only when `cleanup_on_exit` is used?

Reply:
I think it would be preferable to do this regardless of whether cleanup will occur, in order not to leave "dangling" threads in programs embedding the caml runtime (as mentioned in the PR description).

But if people are worried of the consequences of this change, I'll move it to `caml_shutdown`.

Reply:
I don't personally have a strong opinion -- I don't know about threads and systems programming enough to tell the amount of risk involved here -- so I am happy to follow your preference unless someone else speaks out.

Reply:
On the one hand, this adds some time overhead.

On the other hand, I think there is a race in the existing code between `caml_terminate_signals` and the delivery of signals to some other (still-running) domain, which this will fix.